src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -22,13 +36,81 @@ class LeonardoService(gdDAO: DataprocDAO, dbRef: DbReference)(implicit val execu
+    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"
+    for {
+    // Create the firewall rule in the google project if it doesn't already exist, so we can access the cluster
+      _ <- gdDAO.updateFirewallRule(googleProject)
+      // Create the bucket in leo's google bucket and populate with initialization files
+      _ <- initializeBucket(dataprocConfig.leoGoogleBucket, clusterName, bucketName)
+      // Once the bucket is ready, build the cluster
+      clusterResponse <- gdDAO.createCluster(googleProject, clusterName, clusterRequest, bucketName)
+    } yield {
+      clusterResponse
+    }
   }
 
+  /* Create a google bucket and populate it with init files */
+  def initializeBucket(googleProject: GoogleProject, clusterName: String, bucketName: String): Future[Unit] = {
+    for {
+      bucketResponse <- gdDAO.createBucket(googleProject, bucketName)
+      storageObjectsResponse <- initializeBucketObjects(googleProject, clusterName, bucketName)
+    } yield {
+      storageObjectsResponse
+    }
+  }
+
+  /* Process the templated cluster init script and put all initialization files in the init bucket */
+  def initializeBucketObjects(googleProject: GoogleProject, clusterName: String, bucketName: String): Future[Unit] = {
+    val initScriptPath = dataprocConfig.configFolderPath + dataprocConfig.initActionsScriptName
+    val replacements = ClusterInitValues(googleProject, clusterName, bucketName, dataprocConfig).toJson.asJsObject.fields
+
+    template(initScriptPath, replacements).map { content =>
+      val initScriptStorageObject = gdDAO.uploadToBucket(googleProject, bucketName, dataprocConfig.initActionsScriptName, content)
+      // Create an array of all the certs, cluster docker compose file, and the site.conf for the apache proxy
+      val certs = List(dataprocConfig.jupyterServerCrtName, dataprocConfig.jupyterServerKeyName, dataprocConfig.jupyterRootCaPemName,
+        dataprocConfig.clusterDockerComposeName, dataprocConfig.jupyterProxySiteConfName)
+      // Put the rest of the initialization files in the init bucket concurrently
+      val storageObjects = certs.map { certName => gdDAO.uploadToBucket(googleProject, bucketName, certName, new File(dataprocConfig.configFolderPath, certName)) }
+      // Return an array of all the Storage Objects
+    }.recoverWith {
+      case ioError: IOException => throw InitializationFileException(googleProject, clusterName, ioError.getMessage)
+    }
+  }
+
+  /* Process a file using map of replacement values. Each value in the replacement map replaces it's key in the file*/
+  def template(filePath: String, replacementMap: Map[String, JsValue]): Future[String] = {
+    Future {
+      val raw = scala.io.Source.fromFile(filePath).mkString
+      replacementMap.foldLeft(raw)((a, b) => a.replaceAllLiterally("$(" + b._1 + ")", b._2.toString()))
+    }.recoverWith {
+      case t: Throwable => throw TemplatingException(filePath, t.getMessage)
```

Comment:
similarly here. i don't think "wrap all futures in recovers where we turn all possible exceptions into a Leo one" is the right approach -- we should be able to assume this won't throw